doc.go

```diff
@@ -0,0 +1,175 @@
+
+4. If we are above height 0, we need to navigate to the child element at the
+index we calculated, position-mapped via the bitmap. When traversing to the
+child, we discard the upper portion of the index that we no longer need.
+This can be achieved by a mod operation against the number-of-nodes value.
+e.g. an index of 20 at height 1 requires navigation to the element at
+position 2, when moving to that element (which is height 0), we truncate the
+index with 20%8=4, at height 0 this index will be the index in our data
+array (position-mapped via the bitmap).
+
+In this way, each sub-tree root consumes a small slice, log2(width) bits long,
+of the original index.
+
+Adding new elements to an AMT may require up to 3 steps:
+
+1. Increasing the height to accommodate a new index if the current height is not
+sufficient to address the new index. Increasing the height requires turning
+the current root node into an intermediate and adding a new root which
+links to the old (repeated until the required height is reached).
+
+2. Adding any missing intermediate and leaf nodes that are required at the
+index required. There may be large gaps in the AMT which will require
+filling with new nodes.
+
+3. Setting the element at the leaf node in the appropriate position in the data
+array and setting the appropriate bit in the bitmap.
+
+Removing elements requires a reversal of this process. Any empty node (other
+than the case of a completely empty AMT) must be removed and its parent should
+have its child link removed. This removal may recurse up the tree to remove
+many unnecessary intermediate nodes. The root node may also be removed if the
+current height is no longer necessary to contain the range of indexes still in
+the AMT. This can be easily determined if _only_ the first bit of the root's
+bitmap is set, meaning only the left-most is present, which will become the
+new root node (repeated until the new root has more than the first bit set or
+height of 0, the single-node case).
+
+Further Reading
+
+See https://github.com/ipld/specs/blob/master/data-structures/hashmap.md for a
+description of a HAMT algorithm. And
+https://github.com/ipld/specs/blob/master/data-structures/vector.md for a
+description of a similar algorithm to an AMT that doesn't support internal node
+compression and therefore doesn't support sparse arrays.
+
+Usage Considerations
+
+Unlike a HAMT, the AMT algorithm doesn't benefit from randomness introduced by
+a hash algorithm. Therefore an AMT used in cases where user-input can
+influence indexes, larger-than-necessary tree structures may present risks as
+well as the challenge imposed by having a strict upper-limit on the indexes
+addressable by the AMT. A width of 8, using 64-bit integers for indexing,
+allows for a tree height of up to 64/log2(8)=20. Careful placement of indexes
+could be used to force the storage of up to 8^20 elements each in separate leaf
+nodes and the required intermediate nodes to connect them.
```

Comment:
I don't understand that part. I read this as "if you store lots of data, then the AMT will be big". Do you mean that placing indices where each is less than the width apart would lead to a complete AMT, where all intermediate nodes are materialized?

Reply:
yeah, I don't think this is getting the point across so I've tried to rephrase a lot of this bottom section. The main point being that it takes relatively minimal effort if you're a malicious actor who can control indexes to create a _lot_ of nodes in this tree structure that are suboptimal to store and navigate.